File: LeetCode/April Challenge/30th Day/Day_30.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    public:
    bool isValidSequence(TreeNode* root, vector <int> &arr) {
       
        return checkPath(root, arr, 0);
       
    }
   
     private:
     bool checkPath(TreeNode* root, vector <int> &arr, int index)
    {
         if(root == NULL || index == arr.size())
             return false;
        
         if(root->left == NULL && root->right == NULL && root->val == arr[index] && index == arr.size()-1)
             return true;
                  return root->val == arr[index] && (checkPath(root->left, arr, index+1) || checkPath(root->right, arr, index+1));
   
    }
};

File: LeetCode/April Challenge/30th Day/Day_30.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidSequence(self, root: TreeNode, arr: List[int]) -> bool:
        n = len(arr)
        
        def visited(node, index):
            if index == n - 1:
                if node is not None and node.val == arr[index]:
                    return node.left is None and node.right is None
                return False
            if node is None:
                return False
            if node.val == arr[index]:
                return visited(node.left, index + 1) or visited(node.right, index + 1)
            return False
        
        return visited(root, 0)
